### ISSUE TITLE
Multiple prefixes or configs?

### ISSUE THREAD

**Issue by saas786** [humbug/php-scoper#933]
## Feature Request

Suppose I have a dependent project (child theme, plugin, etc.). When the parent theme or main plugin is prefixed, the dependent one isn't, and it definitely fails.

The dilemma is, I can prefix the dependent project, but how do I tell the prefixer that it should perform extra prefixes based on predefined prefixes?

I want the prefixes of the parent project to be passed to the dependent project so that it can learn and apply those prefixes too. This way, when the parent project is prefixed, the dependent project also automatically applies those prefixes.

Upon examining the source code here: https://github.com/humbug/php-scoper/blob/cddf84c08644e0adce018d44daec55b633a87484/src/Console/ConsoleScoper.php#L144, I got the idea that using multiple configs might help. 

Does this make sense?

**Comment by theofidry** [comment]
Hi, sorry I am a bit lost, could you expend a bit on the use case?

**Comment by saas786** [comment]
> Hi, sorry I am a bit lost, could you expend a bit on the use case?

Hmmm, explaining this is a bit tricky :)

# Short explanation

but my main goal is to easily add a prefix to projects that rely on another project. Here's how it works:

    Utilizing Dependencies:
        The prefixed main project contains dependencies.
        Child projects, which may or may not use Composer, need to access these dependencies without installing them separately.

    Case 1 - Child Themes:
	    a) If a child theme doesn't use Composer, I should still be able to automatically adopt the parent theme's prefix during development.
	    b) If a child theme uses Composer, I should not only be able to prefix the child theme, but it needs to use the parent theme's prefix to avoid breaking the code utilizing the parent theme's dependencies.

    Case 2 - Plugins with Themes:
        Similar to Case 1, parent and child themes using a plugin should follow the same logic.

    Case 3 - Modular Composer Framework:
        When the main project is prefixed and it utilizes few of the framework packages, the dependent project's Composer package needs to adopt the main project's prefix.
        Avoid reinstalling packages in the dependent project already present in the main project.

    Overall Approach:
        Dump Composer dependencies and namespaces when the main project is built.
        When a dependent project is prefixed, it should use the parent project's namespaces.
        If the same namespace is used, adopt the parent project's prefix.

This ensures smooth integration and prevents failures when dependent projects are prefixed.

----

# Long explanation

My ultimate goal is to prefix any dependent projects (relying on a prefixed project) so that they can:

- Utilize dependencies from the prefixed project (main or parent) without having to install them into the dependent project, considering the dependent project could have or not have Composer autoloading.

Case 1:
I have a prefixed parent theme that uses Composer for dependencies. Child themes may or may not use Composer.

a) If a child theme doesn't use Composer, I still want to automatically prefix it. This way, I can utilize the parent theme's prefix without manually coding it. During development, I use the original code, and the child theme should automatically adopt the prefix when the parent theme is prefixed.

b) If a child theme uses Composer, I prefix its dependencies. However, as the child theme uses code from the parent theme, and it definitely uses parent theme dependencies, prefixing them separately will cause an issue. The child theme needs to use the parent theme's prefix.

Case 2:
I have a plugin with extensions and addons. Parent and child themes will use this plugin, employing a similar logic to Case 1.

Case 3:
Consider a modular Composer framework (Laravel). The main project uses some packages, while the dependent project utilizes packages not present in the main project. I don't want to reinstall all those packages in the dependent project that are already installed in the main project. When the main project is prefixed, a Composer package in the dependent project (part of the modular Composer framework) is prefixed with the dependent project's prefix, causing a failure. Hence, it needs to be prefixed and be able to use the main project's Composer packages already installed.

Dependent projects (plugins, parent themes, or child themes) may or may not use Composer. They will use code from the main project, which might be utilizing Composer dependencies. When they are prefixed, it might lead to failure.

So, I'm thinking...

When the main project is built, I want a way to dump Composer dependencies, primarily namespaces of Composer packages. I'm unsure if we should dump global classes, etc. When the dependent project is prefixed, it should use the parent project's namespaces. If the same namespace is used in the dependent project (in its code or Composer dependencies), it should adopt the parent project's prefix.

---

I hope this doesn't complicate things further :)

**Comment by theofidry** [comment]
Hi, sorry it took a while to go through it.

I am not sure I completely understand the issue to be honest, but I'll try to answer it.

> I have a prefixed parent theme that uses Composer for dependencies. Child themes may or may not use Composer.

Just to be clear, I really do not think it's a good idea to care about this "detail". Whilst it's inefficient to have the same package scoped twice with different prefixes, any attempt to share it will, most likely, render any work done by PHP-Scoper useless. You would be effectively attempt to re-create what npm <6 was doing, and it was a hot mess despite being a language where scoping things is 1000% easier.

So IMO: if you scope a project, it does not matter that it used Composer or not, it should have the least amount of symbols exposed and it is really advised to not expose symbols of one of its Composer dependency.

>When the dependent project is prefixed, it should use the parent project's namespaces

So there is not really a native way to do this, but you certainly can share the prefix in some way. The `add-prefix` command has a `--prefix` option so you could always pass it around or have a process wrapping your whole scoping process, which could fetch the prefix from the parent project.

<hr>

Another thing to consider I guess is if you do want to share some code between projects. In this case, I think, it is best to properly design this. In other words, you do have a public API, for example it could have a dedicated namespace that you could expose, and the consuming code should be able to access it. Since I assume those projects may be installed separately, hence with incompatible versions, you may want to have some sorts of version checks to fail gracefully

**Comment by theofidry** [comment]
>It would be beneficial if I could use PHP Scoper to dump namespaces for the main project by hooking into the prefixing process and creating a file. This is necessary to avoid any unforeseen consequences, as manually handling namespaces for the main project's Composer dependencies is not ideal. Automation would be preferable.

What exactly do you mean by "dump namespaces"?

>If PHP Scoper could allow me to ignore vendor prefixing for the dependent project that doesn't rely on Composer.

I'm not sure what you mean by that? You should be able to select which files you want to scope already

**Comment by saas786** [comment]
> What exactly do you mean by "dump namespaces"?

I want to extract the namespaces used for project composer dependencies from my sample code mentioned earlier, such as:

- Hybrid\Assets
- Hybrid\Core

This way, I can utilize those namespaces when prefixing a dependent project.

The idea is, similar to `exclude-namespaces`, I could execute `replace-namespaces` by providing `Hybrid\Assets => Dependent_Project_vendor\Hybrid\Assets` for each namespace. That's why I asked if you could provide me with the relevant code where I can implement my custom logic.

---

> I'm not sure what you mean by that? You should be able to select which files you want to scope already

In short, suppose I don't have the Composer vendor in my dependent project. In that case, I just want the prefixing process to proceed with its prefixing without encountering any failures.

**Comment by theofidry** [comment]
>I could execute replace-namespaces by providing Hybrid\Assets => Dependent_Project_vendor\Hybrid\Assets for each namespace. That's why I asked if you could provide me with the relevant code where I can implement my custom logic.

I reckon the gist of it could be done via the patchers: you have the map declared in your config file and you could do a `str_replace()` within your custom patcher leveraging that map.

That said I won't pretend I understood completely your use case, might be easier if you could provide a minimal setup show casing what you're doing

**Comment by saas786** [comment]
> might be easier if you could provide a minimal setup show casing what you're doing

I will try to set up such a configuration as soon as possible, but it may take some time.

**Comment by saas786** [comment]
I have not yet been able to allocate time for testing the repositories. However, I managed to overcome the issue by utilizing `patchers`. Therefore, I am closing this task for now.
